[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113238
* __->__ #112518

This PR has the following goals:
1. Detect unhealthy nccl watchdog thread by implementing a heartbeat. NCCL watchdog sometimes can hang for several reasons such as nccl/cuda API bugs or unexpected blocking behaviors. This is the last resort to ensure that we don't silently keep the training job run for hours.
2. Sometimes, the process gets stuck in the destroy of NCCL PG, and this PR will ensure that we will eventually abort it after some time (by default 2 mins)
3. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
4. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

To clarify further what this PR is trying to solve, we first list are four cases when a NCCL PG can end up with:
- case 1: ncclwatchdog gets stuck (maybe some blocking API) and heartbeat monitor kills it during regular heartbeat monitor loop.
- case 2: ncclwatchdog timeout and desync report or destroy kicked in(let's call it shutdown) but this shutdown takes so long and heartbeat believes it has to kills the process anyway.
- case 3: ncclwatchdog aborts the process (heartbeat monitor not involved)
- case 4: program exits cleanly (heartbeat monitor not involved)

As we can see here, this PR is trying to address case one and two and we also want to ensure adding one more monitor thread does not interfere what we are currently doing in case three and four. That's why we added two flags `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_`.

For case three and four, either `monitorWakeUpCV_` will be waked up in the destructor or `terminateHeartbeatMonitorThread_` will be set to true. So that monitor thread will just exit ASAP.

For case one, both `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_` will still false when monitor thread see there are no heartbeat, so it will directly kill the process. For case two, either `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_` will be true, the monitor thread will wait extra time before killing the process.

cc @seemethere @malfet @osalpekar @atalman

Differential Revision: [D51146305](https://our.internmc.facebook.com/intern/diff/D51146305)

fduwjj committed:
[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,6 +1158,76 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  int counter = -1;
+
+  while (true) {
+    std::unique_lock<std::mutex> lock(heartbeatMutex_);
```
I am not sure if I need to also add try-catch here like watchdog thread does.

wconstab replied:
iiuc this line will hang the watcher thread.  we don't want that

that's the reason that we discussed using pipes to communicate the heartbeat.  In the pipe model (or message Q model) there is no need to share a lock between the two threads.  This lets the 'watcher' thread simply loop/peek at the recv end of the pipe/queue and have no risk of hanging. 

See https://www.bogotobogo.com/cplusplus/multithreading_ipc.php -- you can compare the queue/pipe example with the shared variable example.

fduwjj replied:
wait, then when it calls `wait_for`, it will release the mutex lock right? Let me look into the pipe example. Also cc: @kwen2501

kwen2501 replied:
That's why I thought having a heartbeat variable shared between threads but without a lock paired to it is better than with a lock paired to it. Eventually, as long as the monitoring thread can sense a change in that heartbeat variable, we are good. The read-after-write race is not a concern in this case.

kwen2501 replied:
Oh, if the shared variable is a queue, then it is a whole different story. I was thinking of the shared variable being a `long` counter.

fduwjj replied:
I think we do need a queue here, so do we all agree that we can use a message queue instead? I prefer mq rather than a pipe. @kwen2501 @wconstab

wconstab replied:
if we only care about 'step 1' of your watchdog plan, we could use a shared varaible without a lock.  the variable would be an int that gets incremented by one thread and watched by the other thread.  The 'watcher/other' thread would sleep a certain time, wake, check if the counter increased, if not, raise alarm.

but since we also care about 'step 2', we want a way to decouple the 'watcher' thread out of process. this was where the 'pipe' idea came from.  Now we are publishing a heartbeat message to the pipe from one thread, and seeing it from the other thread.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,6 +1158,76 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+      type_hasher(static_cast<std::uint8_t>(heartBeat.op_type_));
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
```
One easy way to write a test for the watcher is, in c++, acquire the workMetaListMutex_ lock from inside the test code, which would stall the workCleanupLoop.  Then the workCleanupLoop would forget to issue the heartbeat and the test code should expect an exception from the watcher thread

fduwjj replied:
ok, I see.. Yeah, we can start with that as a unit test.

fduwjj committed:
Update on "[WIP][NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 void ProcessGroupNCCL::ncclCommWatchdog() {
   try {
     VLOG(2) << "[Rank " << rank_ << "] NCCL watchdog thread started!";
+    if(monitorThreadEnabled_.load()) {
```
one reason for launching the thread here instead of inside constructor is because it will let the function called by the thread can be overridden.

fduwjj replied:
So to test it in unit test, we want to capture the errors from this thread right? If we create the thread in the constructor of `ProcessGroupNCCL`. Then even if we make `ncclCommsMonitor` virtual and override it in the child class (for testing purpose), it won't work. The thread will still execute `ProcessGroupNCCL::ncclCommsMonitor`, aka the method from parent. That's why I move the thread creation here so that we can do some tricks in test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+          std::chrono::milliseconds(kWatchdogThreadSleepMillis * 3));
+    } else {
+      // Timeout case and we will first dump the flight recorder
+      // to std::out, and we use a different detached thread to ensure
```
i might structure this differently- 
```
else {
  // we clearly exit the loop on first error
 break;
}

// then below we do the 'epilogue' behavior for dumping and crashing
```

also, maybe skip the sideThread for now.  Is dump actually unreliable?  If so we might want a side thread.  (later, if we're dumping to manifold for example).

But in this case, we can probably just call dump, and then ensure we exit/exception

fduwjj replied:
I see. Originally @kwen2501 suggested to use a side thread to keep this thread as clean as possible. For now since we are not doing very complicated thing, I am ok with either way. I am not sure if dumping is unreliable or not. But it's like an extra protection?

wconstab replied:
Let's only add the side thread if we're sure about it.  It otheriwse adds complexity that might get in the way later.

One example is that if we were to launch some RPC calls here, the RPC calls may use their own 'side-thread' so that the calls can return immediately.  In that case we probably don't want our own side-thread.  But if we decide instead to call risky/expensive functions inline we probably should use a side-thread.

However, if we do use a side thread, then we have to consider how long to let it work before we proceed to exit.

Is our plan still that the watchdog stuck will kill the process?  (if so, raising an exception is not good enough, becuase exceptions can call destructors which can then acquire locks and hang).  So maybe we need to call std::abort() at the end of this function to be safe.

Then, I see 2 options
(a) we call dump_debug_info() inside this thread, and then call std::abort().  We must ensure dump_debug_info is a safe function to call.  (if all it does is print some stuff thats in memory, it should not be hard to analyze).  If it needs to check on any more work objects or make additional cuda calls, this would not be good.
(b) use a side thread to call dump_debug_info.   but, then we need to sleep this thread for some time to let the side-thread work, before we finally call abort.

fduwjj replied:
AFAIK, by reading through flight recorder's PR, there seems to be no CUDA call, so it should be ok that we call it here directly. And I see your point, because this is the really last resort, we need to abort the process anyway rather than throwing exceptions.. But that makes unit test hard, there is no way we can do try catch..

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -692,6 +702,16 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   std::mutex mutex_;
 
+  // Heartbeat of watchdog thread.
+  std::atomic<int> heartbeat_;
```
i think we do not want an atomic here.  I am not sure if it is well defined whether an atomic will use a lock or not.  (at best, it may depend on the processor you are compiling to as well as the type (<int>).

We don't really need a lock here.  One thread is always writing, the other thread is always reading.  You may want to look into volatile.

fduwjj replied:
I thought about it, too. Since only one rewriter and one reader, using atomic is an overkill.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ", rank_, "] NCCL monitor thread terminated with timeout: ");
+  monitorException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
```
@malfet had some comments on our pattern of setting the exception this way in our old watchdog code.  I know you're just copying the same pattern, but, i think its probably not the best pattern so would be good to understand why (if there is a reason) or else just do something simpler.

fduwjj replied:
I am more than happy to learn what @malfet might say about this.

fduwjj commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -141,6 +142,36 @@ class ProcessGroupNCCLTimedOutErrors : public ProcessGroupNCCLSimulateErrors {
 
+class ProcessGroupNCCLWatchdogStuckKill : public c10d::ProcessGroupNCCL {
+ public:
+  ProcessGroupNCCLWatchdogStuckKill(
```
yeah this is for unit test. OK, I will rename it.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -630,6 +634,12 @@ class TORCH_API ProcessGroupNCCL : public Backend {
  protected:
+  // Function that runs as part of a separate thread aside from watchdog
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in a collective or watchdog itself hangs,
```
OK, maybe I should rephrase it to "when watchdog gets stuck in NCCL or CUDA call".

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -692,6 +702,16 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+  // Monitor thread which checks the heartbeat of Watchdog thread.
+  // If the monitor thread finds there is no heartbeat, it will dump debug info
+  // and then kill the watchdog thread to avoid hang.
+  std::thread ncclCommMonitorThread_;
```
Are you suggesting that we swap these two threads?

wconstab replied:
similar to @kwen2501's comment.

CommMonitor sounds like a thing that monitors comms.

Watchdog sounds more like a thing that monitors the system.

We currently have the names inverted-- watchdog is monitoring comms, and CommMonitor is monitoring the watchdog.  That is pretty confusing IMO.

I think if you just swap the names its good enough, but we could always bikeshed further

fduwjj replied:
ok, sure. Let me think more between `watchdog` and `watchdog monitor`. Thanks for input from both.

fduwjj replied:
OK, I decide to go with `heartbeatMonitor` and keep the original watchdog for the following reasons:
1. `heartbeatMonitor` the name itself is self-contained in this PR. Down the road, if we add more stuff into this thread, we can always change it later.
2. The name of `watchdog` is widely used and is familiar to users already. So I want to decouple another big change from this PR. (We can do it in the following PR, if we feel the watchdog should be the other around.). Also, today's watchdog is indeed using pull-based way to be "petted" by checking whether there is timeout/exception, etc.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
+  int hearBeatCounter = -1;
```
opps.

fduwjj commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -141,6 +142,36 @@ class ProcessGroupNCCLTimedOutErrors : public ProcessGroupNCCLSimulateErrors {
+    try {
+      ProcessGroupNCCL::ncclCommsMonitor();
+    } catch (std::runtime_error& e) {
+      hasMonitorThreadLaunched_.store(true);
```
OK, the naming may be a little bit confusing here.. like you mentioned below, I will rename it to `HasErrorAndCaughtFlag`

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 }
 
+void dump_debugging_info() {
+  LOG(ERROR) << "Dumped debugging information: " << dump_nccl_trace();
```
Oh, here flight recorder is "information for debugging" in general?

wconstab replied:
I guess what i'm after is something more like

"PGNCCL's watchdog observed <blah>, so it is dumping debug info"

you may also want to point out in the <blah> part that this is a fatal error? or is it not always the case?

fduwjj replied:
oh, I think I misread your comment yesterday. Maybe we want to say something like: "No PGNCCL's watchdog heartbeat detected, we will abort the process after dumping the debug info: " and use `LOG(FATAL)` instead.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  int hearBeatCounter = -1;
+  while (true) {
+    auto heartbeat = heartbeat_.load();
+    if (heartbeat > hearBeatCounter) {
```
Maybe try checking this a couple times (than once) before concluding that there is a timeout? This monitoring thread is not in a hurry and can give the watchdog some "grace period".

fduwjj replied:
Hmmmm as discussed below, we will let the monitoring thread sleep for longer and make it configurable so that we don't need to check multiple times here.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1227,12 +1272,16 @@ void ProcessGroupNCCL::workCleanupLoop() {
       work.checkAndSetException();
       bool timedOut = work.checkTimeout();
+      // Bump up heart beat by one.
+      heartbeat_.fetch_add(1);
```
i mean.. i tend to agree with Ke. But i also understand why @fduwjj did it this way.  It is possible we have a large number of work's and each one takes some amount of time to process.  

Basically, all that i think matters is that we are able to tune the heartbeat frequency adequately.  Example:  we expect to have O(100) works and processing each one takes O(10ms) on average.  Then we could either expect inner-loop heartbeat at 10ms intervals, or outer-loop heartbeat at 1s intervals.  We'd need adequate safety margin on either case.  If inner-loop, we could fail a missed heartbeat more quickly, but, honestly i don't think its critical to detect the stuck watchdog with sub-second accuracy.  Maybe detecting it at minute-level accuracy is enough.  

Otoh, false alarms here will be problematic.  We'd get SEVs if our watchdog is too sensitive and kills normal programs just bc they had more works than expected.  So that also biases me to just set the heartbeat detector to a longer period - e.g. start with several minutes?  

Do we have any data on the actual time spent in the work loop for real cases?  It'd be good to validate this a bit.

wconstab replied:
oh, and I think setting the heartbeat detector to "several minutes" and then keeping the heartbeat inside the inner-loop is also cheating.  IF you want to go that route, you'd have to convince me that it's actually "Expected" for processing one work item to sometimes take a huge amount of time.

This is not any better than multiplying "several minutes" x "a hundred work items on average" and just setting the heartbeat detector to "several hundred minutes" with an outer-loop heartbeat.

fduwjj replied:
> It is possible we have a large number of work's and each one takes some amount of time to process.

That's exactly why I am doing that. And I agree false alarm will make us look really bad. Maybe we can do it in the minute level and make the timeout configurable?

fduwjj replied:
If we set minute-level heartbeat, we don't need the inner heartbeat.

fduwjj replied:
> Do we have any data on the actual time spent in the work loop for real cases? It'd be good to validate this a bit.

We don't...

fduwjj replied:
But for NCCL hang, it's reasonable to assume, we want to kill it after couple minutes?

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  int hearBeatCounter = -1;
+  while (true) {
+    auto heartbeat = heartbeat_.load();
+    if (heartbeat > hearBeatCounter) {
```
ok, maybe let's use unsigned long then? so the range is 0 to 4,294,967,295. If we assume the heartbeat is sent every second, then it is 49K days, roughly 136 years.. That should be sufficient for any types of training jobs.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -630,6 +634,12 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in a collective or watchdog itself hangs,
+  // we can dump the debugging information and abort the process.
+  virtual void ncclCommsMonitor();
```
Sure, maybe I want to think about the naming a little bit.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  const auto exitMsg = c10::str(
+      "[Rank ", rank_, "] NCCL monitor thread terminated with timeout: ");
+  monitorException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
+  if (!terminateProcessGroup_.load()) {
```
We can use that for the while loop as well but here, if the main thread already exited and destroyed, we probably don't want to throw exception again. (Without this check, existing unit test also failes)

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

xw285cornell commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1160,60 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+      heartBeatCounter = heartbeat;
+      // Sleep for enough time so that heartBeat will at least increase.
+      std::this_thread::sleep_for(std::chrono::milliseconds(
+          kWatchdogThreadSleepMillis * heartbeatTimeoutInSec_));
```
it feels the termination will get stuck for up to 2min which is not ideal. I wonder if we should check termination signal more frequently (like having an env var)

fduwjj replied:
This is controlled by env variable `TORCH_NCCL_HEARTBEAT_TIMEOUT` which is in Secs. So if you don't like 2 mins (60 * 2), users can set it to whatever you want.

fduwjj replied:
synced offline, we need to figure out a way to make it wakes up early for the normal case.

wconstab replied:
@xw285cornell what is your expectation? 

We currently live with a 30 min nccl timeout, and try to move people to 10min instead.  Do you think it's important to have sub-minute precision with the heartbeat?

I worry that we don't know enough about the inner-workings of the comm monitor loop to ensure that it gives a reliable heartbeat in a tight bound.

xw285cornell replied:
@wconstab i think the problem is that when a program correctly finished, they may have to wait for up to 2min for the process to clean up which is definitely not good imo

wconstab commented:
> This PR has the following goals:
>1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
>2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
>3. We will then try to throw exceptions to kill the process.

Pls update PR description.  I think its confusing, especially (1).  Detecting a hang in collective is not new functionality in this PR.  This PR tries to detect if the monitoring system is unhealthy.

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -257,3 +297,33 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_HEARTBEAT_TIMEOUT, "3", 1) == 0);
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_ENABLE_MONITORING, "1", 1) == 0);
+  auto options = c10d::ProcessGroupNCCL::Options::create();
+  options->timeout = std::chrono::milliseconds(30000);
```
just a nit- does this test take 30 sec to run? ideally make it faster

wconstab replied:
(generally 1 sec is still 'a long time' and if you dont feel safe maybe use 2 or 3 but not 10?

fduwjj replied:
this is just watchdog timeout it does not affect the time of execution?

fduwjj replied:
Running this test only needs 10s.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1160,60 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // Create a error message reported from MonitorThread, so
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ", rank_, "] NCCL monitor thread terminated with timeout: ");
```
is this incomplete? 

- it looks like you put a `:` and forgot to put the details after it
- it would help to explain more about the timeout.  in this case, it's actually the MonitorThread that timed out not the collective.  (or more precisely, we don't know what caused the MonitorThread to time out).  Could write that this is a system error and please file a bug to pytorch.

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1160,60 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // to std::out. Down the road, if we have more complicated or blocking
+  // operations, we might need to use a side thread to do it.
+  if (parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0) > 0) {
+    // TODO: Find the right and proper way to dump the debug info.
```
oh, I see. sure.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1175,15 +1177,21 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
-      // Sleep for enough time so that heartBeat will at least increase.
-      std::this_thread::sleep_for(std::chrono::milliseconds(
-          kWatchdogThreadSleepMillis * heartbeatTimeoutInSec_));
+    std::unique_lock<std::mutex> lock(monitorMutex_);
```
no locks!

fduwjj replied:
this is fine because this is only used by this thread so that we can wake up early.

fduwjj replied:
Unless we are ok with busy-waiting, like check the flag every 2 secs something like that.

wconstab replied:
i think this approach with lock is ok actually. add a comment to warn people not to acquire it.  
more frequent sleep is probably also fine.

fduwjj replied:
sure, will add comment. I am not a big fan of busy waiting, so I will go with this lock approach.

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1152,9 +1162,69 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+    std::unique_lock<std::mutex> lock(monitorMutex_);
+    if (monitorWakeUpCV_.wait_for(
+            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+              return terminateProcessGroup_.load();
```
we probably need to have another flag, so we keep heartbeat-watcher alive WHILE we shut down the watchdog thread (in case watchdog thread hangs during its own shutdown).

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in a collective call by creating a heartbeat locally.
2. Once detected (suspected), we dump debugging information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to stderr.
3. We will then try to throw exceptions to kill the process.

[ghstack-poisoned]

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -141,6 +142,44 @@ class ProcessGroupNCCLTimedOutErrors : public ProcessGroupNCCLSimulateErrors {
+  // It's really hard to unit test std::abort. So we override it instead.
+  // Commented this override, we do see process aborted with core dump.
+  void terminateProcess(std::string errMsg) override {
+    monitorException_ = std::make_exception_ptr(std::runtime_error(errMsg));
```
why do you do this 'make_exception_ptr' thing and call rethrow?  can't you just `throw runtime_error(errMsg)`?

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -257,3 +297,33 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+    std::lock_guard<std::mutex> lock(pg.getWatchdogMutex());
+    LOG(INFO) << "Lock watchdog thread.";
+    // Wait for a while before monitor thread throws exceptions.
+    std::this_thread::sleep_for(std::chrono::seconds(10));
```
2 suggestions
 - don't hardcode 10 sec here, its better if you compute 2 * heartbeat_interval or something like that
 - maybe for this test case you can even configure the heartbeat interval to be fast, then reduce the test time from 10sec to 1sec.  (just from annoyance, it gets tiring for a test to be slow when you need to repro an issue later)

fduwjj replied:
Well. because `ProcessGroupNCCL::kWatchdogThreadSleepMillis = 1000`, so heartbeat check needs to be at least every 2 secs. That's why I use 10 seconds.. Let me see if I can make it shorter.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1227,6 +1317,8 @@ void ProcessGroupNCCL::workCleanupLoop() {
+    // Bump up heart beat by one.
+    heartbeat_++;
 
     for (auto it = workMetaList_.begin(); it != workMetaList_.end();
```
We probably need to handle the 'desync report' case.

if we go into desync report mode below, desync report may take long enough that heartbeat monitor will kill us.

fduwjj replied:
OK, so I am going to add a another flag for desync report.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1150,11 +1158,93 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+
+  // We need to wait for abort to finish before we can safely shut down
+  // heartbeat monitoring thread.
+  terminateHeartbeatMonitorThread_.store(true);
```
hmm. do we have good enough control over heartbeatMonitor so that it waits long enough after `terminateProcessGroup` is set true, before abort() finishes?

1. What happens inside watchdog thread when we call terminate? 
         Answer: we continue processing a 'for loop' over work objects, and possibly, issue a desync report
3. is the time to terminate + abort expected to be the same as the normal loop period? 
         Answer: not sure how long abort is, but, for-loop over work can be arbitrarily long, so we probably have enough padding in our usual heartbeat period if its safe for for-loop over work, to also account for the abort call

fduwjj replied:
For No.1, we make the heartbeat monitoring wait for long enough so that we keep enough buffer. (We probably need to bump the heartbeat inside desync report generation?)

For No.2, for now we don't really directly terminate inside watchdog, we throw exceptions after generating desync report. (Unless users want to call abort intentionally, like set `NCCL_ASYNC_ERROR_HANDLING` to be 1.)

fduwjj replied:
OK, with two flags added, I am confident that we should be able to capture most corner cases.. And fingers crossed.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -870,6 +872,10 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+  heartbeat_ = 1ULL;
+  monitorThreadEnabled_.store(parseEnvVarFlag(TORCH_NCCL_ENABLE_MONITORING));
+  heartbeatTimeoutInSec_ =
+      parseEnvVarIntDefault(TORCH_NCCL_HEARTBEAT_TIMEOUT, 60 * 2 /*2 Mins*/);
```
nit: good practice to append _S after a variable name that represents time in seconds to avoid confusion.

fduwjj replied:
oh, sure, thanks for the suggestion.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1150,11 +1158,93 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+              return terminateHeartbeatMonitorThread_.load();
+            })) {
+      // If monitorWakeUpCV_ gets notified, we early return.
+      return;
```
By "early return", do you mean the `return` statement here?
Why would we return (exit this function) when there is no timeout (i.e. gets notified)?
I thought we should continue the monitoring.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -797,6 +833,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   std::exception_ptr watchDogException_ = nullptr;
 
+  std::exception_ptr monitorException_ = nullptr;
+
```
This variable is not used in code.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -870,6 +872,10 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   desyncDebug_ = parseEnvVarFlag(NCCL_DESYNC_DEBUG) ||
       (dist_debug_level_ >= DebugLevel::Detail);
+  heartbeat_ = 1ULL;
+  monitorThreadEnabled_.store(parseEnvVarFlag(TORCH_NCCL_ENABLE_MONITORING));
```
We want this to be gradually enrolling instead of turning it on by default. With a gate, we can do more risky or bold things.

fduwjj replied:
If after battle test, this turns out to be a robust feature, we will make it default for sure.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1150,11 +1158,93 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  }
+}
+
+void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+  // Logging with `FATAL`, after errMsg printed, it calls `std::abort()`
+  // to terminate the program execution.
+  LOG(FATAL) << errMsg;
+}
```
yes, that's the purpose and I also mentioned in the comment?

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -6,6 +6,7 @@
 #include <iostream>
 #include <list>
 #include <mutex>
+#include <queue>
```
oh, good catch, let me remove it.

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1226,9 +1227,12 @@ void ProcessGroupNCCL::heartbeatMonitor() {
-    std::this_thread::sleep_for(std::chrono::seconds(120));
+    auto timeout = heartbeatTimeoutInSec_;
+    // Leave six mins for desync report generation or process group destroy.
+    std::this_thread::sleep_for(
```
sigh... 

we're gonna face the same issue Xiaodong mentioned: process-exit is not gonna be fast.  We need to have this sleep also be a CV notifiable one i guess?

wconstab replied:
it might be a simple fix, can you safely use the same CV for this sleep too, and add a .notify call after desync is finished, and also .notify after the ~PGNCCL shutdown sequence finishes and we set the flag to kill watchdog?

fduwjj replied:
OK, I think there are couple scenarios:

1. Everything goes super smooth, abort() goes super fast, we send out notify, and then we directly exit in L1201.
2. The thread happens to wake up naturally and there is no heartbeat because watchdog/main thread exit. So we might need a sleep wake up here.
3. We got stuck in abort.

This sleep is trying to differentiate between 2 and 3. Maybe we can the conditional variable wait here, instead of sleep.

fduwjj replied:
No, but for case #2, `terminateHeartbeatMonitorThread_` is already set to `true`, there is less likely we will enter this sleep. (I cannot 100% exclude it though) Mostly likely when it hit here, it's either we spend too much time in abort or we are generating desync report.

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1150,11 +1158,101 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+      // For the normal complete or user interception, monitorWakeUpCV_
+      // will get notified, we early return and exit heartbeatMonitor.
+      return;
+    } else {
```
nit: this else is not necessary due to the "return". Very minor though, not necessary to change.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1150,11 +1158,101 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  const auto exitMsg = c10::str(
+      "[Rank ",
+      rank_,
+      "] NCCL monitor thread timeout. Basically, this could ",
+      "be a system error and please file a bug to pytorch.");
```
I don't think we can confidently conclude that this is a system error. 
And since this is likely not caused by a PyTorch-internal error (we've done our code right), it would be unnecessary for user to file a bug against PyTorch.

kwen2501 replied:
I think a plainer explanation would suffice, something like:
"ProcessGroupNCCL's watchdog thread is stuck when cleaning or aborting collective work. This can be due to CUDA or NCCL calls being unexpectedly blocking, especially when your program enters a deadlock state. Please refer to PyTorch Distributed's document for investigating the potential deadlock."

wconstab replied:
i think i disagree.  

we should probably write our code such that even if nccl hangs, our code does not hang.  this implies any state where our watchdog hangs "Becuase of nccl" is also our bug.

at this time, are we confident our watchdog can ONLY hang if it's a cuda/nccl bug?

also, do we want to discourage or encourage users filing an issue if there is a hang?  I prefer we know about the problem at least, so we can get it fixed

kwen2501 replied:
This is not any libraries' bug, I'd call it limitation. That's why I said "CUDA or NCCL calls being unexpectedly blocking, especially when your program enters a deadlock state." The cause of deadlock state can be manyfold, such user code mistake, system error, GIL being held somewhere else, etc.

fduwjj replied:
oh, I just mention "due to CUDA or NCCL calls being unexpectedly blocking, especially when your program enters a deadlock state" and remove the system error to avoid confusion. I am open to better suggestions.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1246,7 +1350,10 @@ void ProcessGroupNCCL::workCleanupLoop() {
-            auto desyncMsg = retrieveDesyncReport(store_, "NCCL", rank_, size_);
+            // Set shutdown mode, so the heartbeat monitor thread will not abort
+            // process immediately.
+            shutdownMode_.store(true);
```
nit: how about a clearer name than `shutdownMode`? Such as `gatheringDebugInfo` or `debuggingBeforeExit`.

fduwjj replied:
I see.. Maybe `collectiveDebugInfoMode`?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -44,6 +44,12 @@ constexpr const char* NCCL_DESYNC_DEBUG = "NCCL_DESYNC_DEBUG";
+    "TORCH_NCCL_ENABLE_MONITORING";
+
+constexpr const char* TORCH_NCCL_HEARTBEAT_TIMEOUT_S =
+    "TORCH_NCCL_HEARTBEAT_TIMEOUT_S";
```
nit: S is unnecessary.

wconstab replied:
lol, i asked him to add it. :( 

my concern is that other timeout in PGNCCL is in ms.  maybe this is too ugly but up to you guys

fduwjj replied:
Maybe `TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC`? Well from the perspective of users, we want the environment variable's meaning is intuitive. So it does not cause confusion when they set it and ask questions like "What is the unit of timeout". So I prefer to add extra info here.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -630,6 +636,22 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   void logWorkEnd(WorkNCCL& work);
 
  protected:
+  // Function that runs as part of a separate thread aside from watchdog
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in some NCCL/CUDA calls,
+  // we can dump the debugging information and abort the process.
+  virtual void heartbeatMonitor();
+
+  // Function that directly trigger std::abort so that the whole process
+  // gets terminated.
+  virtual void terminateProcess(std::string errMsg);
```
nit: should these two functions be more like private than protected?

fduwjj replied:
We cannot override a private method... I agree with you but this is c++, so we cannot do things like in Python.

kwen2501 reviewed (approved):
Approving. 
There are some minor comments/personal preferences for reference.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1150,11 +1158,101 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+      return;
+    } else {
+      auto heartbeat = heartbeat_;
+      if (heartbeat > heartBeatCounter) {
```
i also picked on this before. junjie made the int bigger and claimed we'll never hit it.  I prefer we handle the wraparound too.  (but if we do the wraparound wrong, its worse than if we don't do it at all, and i think it could be  a bit complicated to do the wraparound).

fduwjj replied:
`heartbeat_` is unsigned long, so quote my reply to one of Will's previous comment:

> so the range is 0 to 4,294,967,295. If we assume the heartbeat is sent every second, then it is 49K days, roughly 136 years.. That should be sufficient for any types of training jobs.

fduwjj replied:
OK, I think I will just change `>` to `!=` since heartBeat can be assumed monotonically increasing.

fduwjj replied:
And we will keep unsigned long for now.

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -257,3 +325,75 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+  ASSERT_TRUE(setenv(c10d::NCCL_DESYNC_DEBUG, "1", 1) == 0);
+  // We cannot capture the exception thrown in watchdog thread without making
+  // lots of changes to the code. So we don't let the watchdog throw exception.
+  ASSERT_TRUE(setenv(c10d::NCCL_ASYNC_ERROR_HANDLING, "0", 1) == 0);
```
didn't understand why we have to set async = 0 to do this test.

what if we leave async=1, then let a nccl timeout happen.  (issue collective on rank0 not rank1).  then we 'observe' that desync report gets run (bc we made a class function that we can override in the test).  then we check that heartbeat monitor thread shuts down cleanly (without firing).

wconstab replied:
then also, we can add an alternate version of this test where everything is the same as above, except we make desyncDebug function hang for too long (sleep).  We test that heartbeat monitor fires and kills in this case.

fduwjj replied:
setting `NCCL_ASYNC_ERROR_HANDLING ` to 0 is to avoid watchdog throw exception and get the whole process aborted which cannot be try-catched. I can add logic of whether monitor filed or not.. And also will add one more test case which mentioned in your second comment.

wconstab replied:
ok i think this is fine.

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51146305).

fduwjj commented:
Rebase and import into internal to see CI signals.

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

Differential Revision: [D51146305](https://our.internmc.facebook.com/intern/diff/D51146305)

[ghstack-poisoned]

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect the potential hang/stuck in the watchdog thread (or calling into abort) during PG destroy by listening to a heartbeat locally. The former is to detect if the NCCL watchdog system is still healthy and the latter is to ensure we don't silently let the process hang there for hours when calling Cuda APIs.
2. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
3. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

Differential Revision: [D51146305](https://our.internmc.facebook.com/intern/diff/D51146305)

[ghstack-poisoned]

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -257,3 +345,112 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+
+class ProcessGroupNCCLWatchdogTimeoutTest : public ProcessGroupNCCLErrorsTest {
+ protected:
+  void watchdogTimeoutTestStepup() {
```
nit: is this supposed to be named 'Setup' instead of 'Stepup'?

fduwjj replied:
opps, type.. My bad

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -257,3 +345,112 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
+  c10::intrusive_ptr<c10d::ProcessGroupNCCL::Options> options_;
+};
+
+TEST_F(ProcessGroupNCCLWatchdogTimeoutTest, testNCCLTimedoutDebugInfoFinished) {
```
new tests look good to me. thx for adding!

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1256,11 +1264,104 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // Case two: desync might be slow or get stuck. Or we get stuck in
+  // destructors, we will sleep for some time before calling std::abort() to
+  // kill the whole process.
+  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load()) &&
```
maybe we should add additional logs (info level? warning?) that let us know which path we took in the heartbeat shutdown.

I am thinking of a case where the heartbeat kills the process but we don't know from the logs whether it killed during the  work-monitoring loop being stuck, or during the shutdown

wconstab replied:
if we do add log, it  might be good to include all the various state variables such as which flags are set for watchdog/monitor shutdown/error

fduwjj replied:
Sure, let me add the loggings.

kwen2501 reviewed (approved):
Thanks much for the iterations. LGTM!

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect unhealthy nccl watchdog thread by implementing a heartbeat. NCCL watchdog sometimes can hang for several reasons such as nccl/cuda API bugs or unexpected blocking behaviors. This is the last resort to ensure that we don't silently keep the training job run for hours.
2. Sometimes, the process gets stuck in the destroy of NCCL PG, and this PR will ensure that we will eventually abort it after some time (by default 2 mins)
3. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
4. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

cc seemethere malfet osalpekar atalman

Differential Revision: [D51146305](https://our.internmc.facebook.com/intern/diff/D51146305)

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -987,24 +993,24 @@ ProcessGroupNCCL::ProcessGroupNCCL(
-            << options_->is_high_priority_stream
-            << ", TORCH_DISTRIBUTED_DEBUG: "
-            << std::string(torch_distributed_debug)
+  LOG(INFO)
+      << "[Rank " << rank_ << "] ProcessGroupNCCL initialization options: "
+      << "NCCL version: " << getNcclVersion()
+      << ", NCCL_ASYNC_ERROR_HANDLING: " << asyncErrorHandling_
+      << ", NCCL_DESYNC_DEBUG: " << desyncDebug_
+      << ", NCCL_ENABLE_TIMING: " << enableTiming_.load()
+      << ", NCCL_BLOCKING_WAIT: " << blockingWait_
+      << ", TIMEOUT(ms): " << options_->timeout.count()
+      << ", USE_HIGH_PRIORITY_STREAM: " << options_->is_high_priority_stream
+      << ", TORCH_DISTRIBUTED_DEBUG: " << std::string(torch_distributed_debug)
```
These are moved by `lintrunner`

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -987,24 +993,24 @@ ProcessGroupNCCL::ProcessGroupNCCL(
 #endif
-            << ", NCCL_DEBUG: " << std::string(nccl_debug)
-            << ", ID=" << this->getID();
+      << ", TORCH_NCCL_ENABLE_MONITORING: " << monitorThreadEnabled_.load()
+      << ", TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC: " << heartbeatTimeoutInSec_
```
Added loggings for the flag here. cc: @wconstab

wconstab replied:
cool. this actually wasn't my point, but its still good to add.

(my point was more related to differentiating between case 1 and case 2 in the comment below about issuing a warning on heartbeat kill.  i thought that certain state vars might help distinguish the 2 cases)

fduwjj replied:
That's all NCCL PG flags being printed so I think we should also add loggings here.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1256,11 +1264,110 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+  // kill the whole process.
+  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load()) &&
+      !terminateHeartbeatMonitorThread_.load()) {
+    const auto logMsg = c10::str(
+        "[Rank ",
+        rank_,
+        "] NCCL watchdog is collecting debug info or process is being terminated",
+        "heartbeat monitor thread sleep for some time before abort the process.");
+    LOG(INFO) << logMsg;
```
With this log, we can differentiate between "watchdog being stuck" and "stuck during the shutdown or desync report".

wconstab replied:
by putting the LOG here before the sleep, its possible we print it in a case when nothing bad actually happens.  This is more of a spammy case.

IMO this should be a warning, but, it should only be printed if the heartbeat monitor actually killed the process.

case 1: heartbeat monitor kills during regular heartbeat monitor loop
case 2: heartbeat monitor kills during shutdown period
case 3: ncclwatchdog aborts, heartbeat monitor not involved
case 4: program exits cleanly

what i'm proposing is that only in case 1 and 2, we issue a warning.  But there is enough information in the warning that we can tell whether its case 1 or case 2.

for case 3, we do not want heartbeat monitor to print anything, its just noisy/distracting and not helpful.  similary for case 4.

fduwjj replied:
that's why this log can help differentiate between case 1 and case 2? I do agree we shouldn't print anything for case 3 and case 4.

First of all, for case 3 and case 4, it will not enter both if statements because `terminateHeartbeatMonitorThread_` is set to true very quickly.

For case 1, it will not hit the first if statement because watchdog somehow gets stuck, `collectiveDebugInfoMode_` will be false and `terminateProcessGroup_` is false as well.

For case 2, either `collectiveDebugInfoMode_` or `terminateProcessGroup_` will be true. That's why I added a flag here.

Both case 1 and 2, we will see an abort initiated from monitoring thread and a coredump. Does this sound make sense to you?

fduwjj replied:
Updated the how the flags work in different scenarios in the description and also I just borrowed your languages of case 1-4. So that everyone who read this PR later will understand our thinking process.

wconstab replied:
i see, ok i think im mostly convinced.

> First of all, for case 3 and case 4, it will not enter both if statements because terminateHeartbeatMonitorThread_ is set to true very quickly.

maybe there is technically a race condition for this case though? (is it possible we print the log first and then later the  `terminateHeartbeatMonitorThread_` value is set?

fduwjj replied:
oh, that is possible but very unlikely. We set `terminateHeartbeatMonitorThread_ ` before monitorWakeUpCV_.notify_one(); And we will sleep for two minutes. The only race condition might be, monitor thread happens to wake up and then it finds that no heartbeat and then hit this log.. OK, I think what I can do is to move this logging to the second if and print out the flags. Then we can avoid the race condition.

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51146305).

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect unhealthy nccl watchdog thread by implementing a heartbeat. NCCL watchdog sometimes can hang for several reasons such as nccl/cuda API bugs or unexpected blocking behaviors. This is the last resort to ensure that we don't silently keep the training job run for hours.
2. Sometimes, the process gets stuck in the destroy of NCCL PG, and this PR will ensure that we will eventually abort it after some time (by default 2 mins)
3. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
4. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

To clarify further what this PR is trying to solve, we first list are four cases when a NCCL PG can end up with:
- case 1: ncclwatchdog gets stuck (maybe some blocking API) and heartbeat monitor kills it during regular heartbeat monitor loop.
- case 2: ncclwatchdog timeout and desync report or destroy kicked in(let's call it shutdown) but this shutdown takes so long and heartbeat believes it has to kills the process anyway.
- case 3: ncclwatchdog aborts the process (heartbeat monitor not involved)
- case 4: program exits cleanly (heartbeat monitor not involved)

As we can see here, this PR is trying to address case one and two and we also want to ensure adding one more monitor thread does not interfere what we are currently doing in case three and four. That's why we added two flags `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_`.

For case three and four, either `monitorWakeUpCV_` will be waked up in the destructor or `terminateHeartbeatMonitorThread_` will be set to true. So that monitor thread will just exit ASAP.

For case one, both `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_` will still false when monitor thread see there are no heartbeat, so it will directly kill the process. For case two, either `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_` will be true, the monitor thread will wait extra time before killing the process.

cc seemethere malfet osalpekar atalman

Differential Revision: [D51146305](https://our.internmc.facebook.com/intern/diff/D51146305)

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51146305).

fduwjj committed:
Update on "[NCCL PG] ADD a separate monitoring thread to ensure we collect debug info and check watchdog heartbeat"


This PR has the following goals:
1. Detect unhealthy nccl watchdog thread by implementing a heartbeat. NCCL watchdog sometimes can hang for several reasons such as nccl/cuda API bugs or unexpected blocking behaviors. This is the last resort to ensure that we don't silently keep the training job run for hours.
2. Sometimes, the process gets stuck in the destroy of NCCL PG, and this PR will ensure that we will eventually abort it after some time (by default 2 mins)
3. Once heartbeat cannot be heard, we dump debug information (for now, we just use the flight recorder implemented in https://github.com/pytorch/pytorch/pull/110960/files) to disk. (How and where to dump the debug info will be addressed in the following PR).
4. Finally, we initiate std::abort via `LOG(FATAL)` to kill the process.

To clarify further what this PR is trying to solve, we first list are four cases when a NCCL PG can end up with:
- case 1: ncclwatchdog gets stuck (maybe some blocking API) and heartbeat monitor kills it during regular heartbeat monitor loop.
- case 2: ncclwatchdog timeout and desync report or destroy kicked in(let's call it shutdown) but this shutdown takes so long and heartbeat believes it has to kills the process anyway.
- case 3: ncclwatchdog aborts the process (heartbeat monitor not involved)
- case 4: program exits cleanly (heartbeat monitor not involved)

As we can see here, this PR is trying to address case one and two and we also want to ensure adding one more monitor thread does not interfere what we are currently doing in case three and four. That's why we added two flags `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_`.

For case three and four, either `monitorWakeUpCV_` will be waked up in the destructor or `terminateHeartbeatMonitorThread_` will be set to true. So that monitor thread will just exit ASAP.

For case one, both `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_` will still false when monitor thread see there are no heartbeat, so it will directly kill the process. For case two, either `terminateHeartbeatMonitorThread_` and `collectiveDebugInfoMode_` will be true, the monitor thread will wait extra time before killing the process.

cc seemethere malfet osalpekar atalman

Differential Revision: [D51146305](https://our.internmc.facebook.com/intern/diff/D51146305)

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51146305).

wconstab reviewed (approved):
Thanks for getting this done @fduwjj!

fduwjj commented:
@pytorchbot merge
